### PR TITLE
Snapshot catalog reader and configuration files

### PR DESCRIPTION
**Pull request by evevkovacs**
Here's a DESCQA test run on this catalog: (https://portal.nersc.gov/project/lsst/descqa/v2/?run=2019-04-23_1&catalog=baseDC2_snapshot_z1.01_v0.1)
I made a minimal set of modifications to the code so that I could use the cosmoDC2 parent class.

### CODE REVIEW

**Comment by yymao** [comment]
Thanks, @evevkovacs. The changes look reasonable. I have some minor questions regarding the step/redshift variables that I'll ask during my review.

In the meantime, since the user-facing part is basically ready, I think we can already try to find someone to use this PR to write a tutorial as @rmandelb suggested.

**Comment by evevkovacs** [comment]
@yymao  Thanks Yao. Someone in TJP might be interested in getting started on this.

**Comment by yymao** [comment]
Also, should we avoid using `#` in filenames? I know it is a valid character but it may cause some unexpected annoyance?

**Comment by evevkovacs** [comment]
No there is no iteration over step or redshift, so it would be fine to change the yaml file as you suggested. The # character exists in our filenames for the halo catalogs on which the snapshots are based and have caused no trouble. They refer to the block number in our accounting so we want to retain them. They provide spatially group halos.

File: GCRCatalogs/cosmodc2.py [hunk 1]
```diff
@@ -229,13 +282,21 @@ def _check_cosmology(self, fh, file_name, atol):
...
+        if lightcone:
+            sky_area = dict()
+            max_healpixel = max(hpx_this for _, hpx_this in self._healpix_files)
+            min_valid_nside = hp.pixelfunc.get_min_valid_nside(max_healpixel)
+            default_sky_area = hp.nside2pixarea(min_valid_nside, degrees=True)
+            files_to_check = self._healpix_files
+        else:
+            default_box_size = 3000.
+            box_size = 0.
+            files_to_check = self._snapshot_files
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
```

File: GCRCatalogs/cosmodc2.py [hunk 2]
```diff
@@ -229,13 +282,21 @@ def _check_cosmology(self, fh, file_name, atol):
...
-        min_valid_nside = hp.pixelfunc.get_min_valid_nside(max_healpixel)
-        default_sky_area = hp.nside2pixarea(min_valid_nside, degrees=True)
+        lightcone = self.lightcone
+
+        if lightcone:
+            sky_area = dict()
+            max_healpixel = max(hpx_this for _, hpx_this in self._healpix_files)
+            min_valid_nside = hp.pixelfunc.get_min_valid_nside(max_healpixel)
+            default_sky_area = hp.nside2pixarea(min_valid_nside, degrees=True)
+            files_to_check = self._healpix_files
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
```

File: GCRCatalogs/cosmodc2.py [hunk 3]
```diff
@@ -245,7 +306,7 @@ def _process_metadata(self, ensure_quantity_consistent=False, # pylint: disable=
             check_md5 = False
             warnings.warn('Not able to perform md5 check: no md5 sum specified in {}'.format(CHECK_FILE_PATH))
 
-        for (_, hpx_this), file_path in self._healpix_files.items():
+        for (_, file_key), file_path in files_to_check.items():
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
        for file_key, file_path in self._file_list.items():
```

File: GCRCatalogs/cosmodc2.py [hunk 4]
```diff
@@ -261,13 +322,25 @@ def _process_metadata(self, ensure_quantity_consistent=False, # pylint: disable=
...
-                    sky_area_this = default_sky_area
-                if sky_area.get(hpx_this, 0) < sky_area_this:
-                    sky_area[hpx_this] = sky_area_this
+                if lightcone:
+                    # get sky area
+                    try:
+                        sky_area_this = float(fh['metaData/skyArea'].value) # pylint: disable=E1101
+                    except KeyError:
+                        sky_area_this = default_sky_area
+                    if sky_area.get(file_key, 0) < sky_area_this:
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
                    hpx_this = file_key[1]
                    if sky_area.get(hpx_this, 0) < sky_area_this:
```

File: GCRCatalogs/cosmodc2.py [hunk 5]
```diff
@@ -261,13 +322,25 @@ def _process_metadata(self, ensure_quantity_consistent=False, # pylint: disable=
...
+                        sky_area[file_key] = sky_area_this
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
                        sky_area[hpx_this] = sky_area_this
```

File: GCRCatalogs/cosmodc2.py [hunk 6]
```diff
@@ -276,19 +349,37 @@ def _process_metadata(self, ensure_quantity_consistent=False, # pylint: disable=
...
+        else:
+            for (step_this, block_this), file_path in self._snapshot_files.items():
+                d = {'block': block_this}
+                if native_filters is not None and not native_filters.check_scalar(d):
+                    continue
+                with h5py.File(file_path, 'r') as fh:
+                    for group in self._get_group_names(fh):
+                        if len(fh[group]):
+                            yield lambda native_quantity: fh['{}/{}'.format(group, native_quantity)].value
+
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]: Lines 361-382
```python
        if self.lightcone:
			key_to_dict = lambda key: dict(zip(('redshift_block_lower', 'healpix_pixel'), key))
		else:
            key_to_dict = lambda key: {'block': key}
        for key, file_path in self._file_list.items():
            d = key_to_dict(key)
            if native_filters is not None and not native_filters.check_scalar(d):
                continue
            with h5py.File(file_path, 'r') as fh:
                for group in self._get_group_names(fh):
                    # pylint: disable=E1101,W0640
                    if len(fh[group]):
                        yield lambda native_quantity: fh['{}/{}'.format(group, native_quantity)].value
```

File: GCRCatalogs/cosmodc2.py [hunk 7]
```diff
@@ -261,13 +322,25 @@ def _process_metadata(self, ensure_quantity_consistent=False, # pylint: disable=
...
+                else:
+                    try:
+                        box_size_this = fh['metaData/box_size'].value
+                    except KeyError:
+                        box_size_this = default_box_size
+                    box_size = max(box_size, box_size_this)
+                    try:
+                        self.redshift = fh['metaData/redshift'].value
+                    except KeyError:
+                        self.redshift = self._redshift
```
Comment (yymao) [inline GCRCatalogs/cosmodc2.py]:
```suggestion
```

**Comment by yymao** [comment]
Thanks, @evevkovacs. I've created a PR to update this PR, for you to review: https://github.com/evevkovacs/gcr-catalogs/pull/1.

**Comment by evevkovacs** [comment]
I approved your changes and left a comment.

**Comment by evevkovacs** [comment]
@yymao  OK, I fixed the quantity modifier for baseDC2. It's tested and I think ready to merge.

**Pull request merged**